ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -576,31 +581,11 @@ private boolean isExpired(IndexValue value) {
   FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws StoreException {
     long startTimeInMs = time.milliseconds();
     try {
-      boolean tokenWasReset = false;
+      boolean tokenWasReset;
       Offset logEndOffsetBeforeFind = log.getEndOffset();
-      StoreFindToken storeToken = (StoreFindToken) token;
-      // validate token
-      if (storeToken.getSessionId() == null || storeToken.getSessionId().compareTo(sessionId) != 0) {
-        // the session has changed. check if we had an unclean shutdown on startup
-        if (!cleanShutdown) {
-          // if we had an unclean shutdown and the token offset is larger than the logEndOffsetOnStartup
-          // we reset the token to logEndOffsetOnStartup
-          if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized)
-              && storeToken.getOffset().compareTo(logEndOffsetOnStartup) > 0) {
-            logger.info("Index : " + dataDir + " resetting offset after not clean shutdown " + logEndOffsetOnStartup
-                + " before offset " + storeToken.getOffset());
-            storeToken = new StoreFindToken(logEndOffsetOnStartup, sessionId);
-            tokenWasReset = true;
-          }
-        } else if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized)
-            && storeToken.getOffset().compareTo(logEndOffsetOnStartup) > 0) {
-          logger.error(
-              "Index : " + dataDir + " invalid token. Provided offset is outside the log range after clean shutdown");
-          // if the shutdown was clean, the offset should always be lesser or equal to the logEndOffsetOnStartup
-          throw new IllegalArgumentException(
-              "Invalid token. Provided offset is outside the log range after clean shutdown");
-        }
-      }
+      AtomicReference<StoreFindToken> storeFindTokenAtomicReference = new AtomicReference<>((StoreFindToken) token);
```

Comment:
while this is a perfectly valid way of resetting, I'm wondering if the `AtomicReference` is overkill. What if the the function returned a `StoreToken` and you compared equality? Something like
```
StoreFindToken unprocessedToken = (StoreFindToken) token;
StoreFindToken storeToken = processToken(unprocessedToken);
boolean inclusive = !unprocessedToken.equals(storeToken);
```

Also, you can get rid of the `tokenWasReset` variable since it is only used to set `inclusive` later.

@pnarayanan - Could you also comment on this?